### PR TITLE
[GEP-1911] h2c backend protocol conformance

### PR DESCRIPTION
**Pull request by dprotaso**
/kind test
/area conformance

**What this PR does / why we need it**:

This PR adds a conformance test validating implementations support h2c when the target Kubernetes Service Service Port has a `kubernetes.io/h2c` `appProtocol`

Depends on echo-basic changes which are being bumped here: https://github.com/kubernetes-sigs/gateway-api/pull/2456

**Which issue(s) this PR fixes**:

Related https://github.com/kubernetes-sigs/gateway-api/issues/1911

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by shaneutt** [comment]
Marking draft while marked "WIP" as that helps us to organize reviews a little better.

**Comment by robscott** [comment]
Thanks @dprotaso! This looks reasonable to me, but will defer to conformance approvers on this one.

/cc @arkodg @sunjayBhatia @mlavacca

File: conformance/tests/httproute-backend-protocol-h2c.go
Comment (arkodg) [inline conformance/tests/httproute-backend-protocol-h2c.go:51]: ha good to know

**Review by arkodg — APPROVED** [review]
looks good !
/approve

**Review by sunjayBhatia — COMMENTED** [review]
looks like https://github.com/kubernetes-sigs/gateway-api/blob/0cc3e0e7e8b1660b72340893a80e1464df095435/conformance/conformance_test.go#L49-L50 needs to add the v1 api as well, otherwise tests don't run (same for main so not exclusive to this PR)

**Review by sunjayBhatia — APPROVED** [review]
other than the previous comment, everything seems to work, (I cheated a little with Contour and added the required annotation contour supports for h2c to get the tests to pass/fail accordingly since we don't support the appprotocol field yet)

**Review by sunjayBhatia — CHANGES_REQUESTED** [review]
I guess I'll request changes officially still since I have some outstanding requests and there is a prerequisite change that needs to happen before these tests actually run successfully

File: conformance/base/manifests.yaml
Comment (dprotaso) [reply conformance/base/manifests.yaml:94]: So I don't want to put the protocol name here. It might inadvertedly inform the implementation what the backend protocol is.

eg. istio if you put http2 it will assume h2c

**Comment by dprotaso** [comment]
> I guess I'll request changes officially still since I have some outstanding requests and there is a prerequisite change that needs to happen before these tests actually run successfully

what do you mean?

**Comment by sunjayBhatia** [comment]
> > I guess I'll request changes officially still since I have some outstanding requests and there is a prerequisite change that needs to happen before these tests actually run successfully
> 
> what do you mean?

https://github.com/kubernetes-sigs/gateway-api/pull/2456#pullrequestreview-1690743992

the v1 api group/version needs to be added to the scheme otherwise the tests don't run

otherwise everything looks good

File: conformance/base/manifests.yaml
Comment (sunjayBhatia) [reply conformance/base/manifests.yaml:94]: oh, well then can be disregarded

**Comment by dprotaso** [comment]
Made a separate PR: https://github.com/kubernetes-sigs/gateway-api/pull/2511

File: conformance/utils/suite/features.go
Comment (mlavacca) [inline conformance/utils/suite/features.go:137]:
```suggestion
	// This option indicates support for HTTPRoute with a backendref with an appProtocol 'kubernetes.io/h2c'
```

**Comment by mlavacca** [comment]
Since @sunjayBhatia's requested changes have been addressed in a separate PR, I think we can merge this one, as no other concern popped up.

/lgtm

**Pull request merged**
